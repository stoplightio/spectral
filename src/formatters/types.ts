import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
