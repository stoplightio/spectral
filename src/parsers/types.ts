import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
