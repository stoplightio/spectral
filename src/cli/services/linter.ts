import { isAbsolute, resolve } from '@stoplight/path';
import { IParserResult, Optional } from '@stoplight/types';
import { getLocationForJsonPath, parseWithPointers } from '@stoplight/yaml';

import {
  isJSONSchema,
  isJSONSchemaDraft2019_09,
  isJSONSchemaDraft4,
  isJSONSchemaDraft6,
  isJSONSchemaDraft7,
  isJSONSchemaLoose,
  isOpenApiv2,
  isOpenApiv3,
} from '../../formats';
import { readParsable } from '../../fs/reader';
import { httpAndFileResolver } from '../../resolvers/http-and-file';
import { getDefaultRulesetFile } from '../../rulesets/loader';
import { readRuleset } from '../../rulesets/reader';
import { isRuleEnabled } from '../../runner';
import { Spectral } from '../../spectral';
import { FormatLookup, IParsedResult, RuleCollection } from '../../types';
import { ILintConfig } from '../../types/config';
import { IRuleset } from '../../types/ruleset';

export async function loadRulesets(cwd: string, rulesetFiles: string[]): Promise<IRuleset> {
  if (rulesetFiles.length === 0) {
    return {
      functions: {},
      rules: {},
    };
  }

  return readRuleset(rulesetFiles.map(file => (isAbsolute(file) ? file : resolve(cwd, file))));
}

const KNOWN_FORMATS: Array<[string, FormatLookup, string]> = [
  ['oas2', isOpenApiv2, 'OpenAPI 2.0 (Swagger) detected'],
  ['oas3', isOpenApiv3, 'OpenAPI 3.x detected'],
  ['json-schema', isJSONSchema, 'JSON Schema detected'],
  ['json-schema-loose', isJSONSchemaLoose, 'JSON Schema (loose) detected'],
  ['json-schema-draft4', isJSONSchemaDraft4, 'JSON Schema Draft 4 detected'],
  ['json-schema-draft6', isJSONSchemaDraft6, 'JSON Schema Draft 6 detected'],
  ['json-schema-draft7', isJSONSchemaDraft7, 'JSON Schema Draft 7 detected'],
  ['json-schema-2019-09', isJSONSchemaDraft2019_09, 'JSON Schema Draft 2019-09 detected'],
];

export async function lint(name: string, flags: ILintConfig, rulesetFile: Optional<string[]>) {
  if (flags.verbose) {
    console.info(`Linting ${name}`);
  }

  const targetUri = !/^https?:\/\//.test(name) && !isAbsolute(name) ? resolve(process.cwd(), name) : name;

  const spec: IParserResult = parseWithPointers(await readParsable(targetUri, { encoding: flags.encoding }), {
    ignoreDuplicateKeys: false,
    mergeKeys: true,
  });

  const rulesetFiles = rulesetFile || (await getDefaultRulesetFile(process.cwd()));
  const ruleset = await (rulesetFiles
    ? loadRulesets(process.cwd(), Array.isArray(rulesetFiles) ? rulesetFiles : [rulesetFiles])
    : readRuleset('spectral:oas'));

  const spectral = new Spectral({ resolver: httpAndFileResolver });

  for (const [format, lookup, message] of KNOWN_FORMATS) {
    spectral.registerFormat(format, document => {
      if (lookup(document)) {
        if (!flags.quiet) {
          console.log(message);
        }

        return true;
      }
      return false;
    });
  }

  spectral.setRuleset(ruleset);

  if (flags.verbose) {
    if (ruleset) {
      const rules = Object.values(spectral.rules);
      console.info(`Found ${rules.length} rules (${rules.filter(isRuleEnabled).length} enabled)`);
    } else {
      console.info('No rules loaded, attempting to detect document type');
    }
  }

  if (flags.skipRule) {
    spectral.setRules(skipRules(ruleset.rules, flags));
  }

  const parsedResult: IParsedResult = {
    source: targetUri,
    parsed: spec,
    getLocationForJsonPath,
  };

  return spectral.run(parsedResult, {
    resolve: {
      documentUri: targetUri,
    },
  });
}

const skipRules = (rules: RuleCollection, flags: ILintConfig): RuleCollection => {
  const skippedRules: string[] = [];
  const invalidRules: string[] = [];

  if (flags.skipRule !== undefined) {
    for (const rule of Array.isArray(flags.skipRule) ? flags.skipRule : [flags.skipRule]) {
      if (rule in rules) {
        delete rules[rule];
        skippedRules.push(rule);
      } else {
        invalidRules.push(rule);
      }
    }
  }

  if (invalidRules.length !== 0 && !flags.quiet) {
    console.warn(`ignoring invalid ${invalidRules.length > 1 ? 'rules' : 'rule'} "${invalidRules.join(', ')}"`);
  }

  if (skippedRules.length !== 0 && flags.verbose) {
    console.info(`skipping ${skippedRules.length > 1 ? 'rules' : 'rule'} "${skippedRules.join(', ')}"`);
  }

  return rules;
};
