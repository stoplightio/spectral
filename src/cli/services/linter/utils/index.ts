export * from './getRuleset';
export * from './listFiles';
export * from './skipRules';
