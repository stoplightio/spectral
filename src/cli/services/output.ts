import { Dictionary } from '@stoplight/types';
import { writeFile } from 'fs';
import { promisify } from 'util';
import { html, json, junit, stylish, teamcity, text } from '../formatters';
import { Formatter, FormatterOptions } from '../formatters/types';
import { IRuleResult } from '../../types';
import { OutputFormat } from '../../types/config';

const writeFileAsync = promisify(writeFile);

const formatters: Dictionary<Formatter, OutputFormat> = {
  json,
  stylish,
  junit,
  html,
  text,
  teamcity,
};

export function formatOutput(results: IRuleResult[], format: OutputFormat, formatOptions: FormatterOptions): string {
  return formatters[format](results, formatOptions);
}

export async function writeOutput(outputStr: string, outputFile?: string) {
  if (outputFile) {
    return writeFileAsync(outputFile, outputStr);
  }
  console.log(outputStr);
}
