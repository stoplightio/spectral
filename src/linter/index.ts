import { PathComponent } from 'jsonpath';
import { filter, omitBy } from 'lodash';
import { IRule, IRuleOpts, IRuleResult } from '../types';
import { IRuleEntry, IRunOpts } from '../types/spectral';
const get = require('lodash/get');
const has = require('lodash/has');

// TODO(SO-23): unit test but mock whatShouldBeLinted
export function lintNode(
  ruleEntry: IRuleEntry,
  opts: IRunOpts,
  node: { path: PathComponent[]; value: any }
): IRuleResult[] {
  const conditioning = whatShouldBeLinted(node.value, ruleEntry.rule);
  // If the 'when' condition is not satisfied, simply don't run the linter
  if (!conditioning.lint) {
    return [];
  }

  const opt: IRuleOpts = {
    object: conditioning.value,
    rule: ruleEntry.rule,
  };

  if (ruleEntry.rule.given === '$') {
    // allow resolved and stringified targets to be passed to rules when operating on
    // the root path
    if (opts.resolvedTarget) {
      opt.resObj = opts.resolvedTarget;
    }
  }

  return ruleEntry.apply(opt, {
    given: node.path,
  });
}

// TODO(SO-23): unit test idividually
export function whatShouldBeLinted(originalValue: any, rule: IRule): { lint: boolean; value: any } {
  const when = rule.when;
  if (!when) {
    return {
      lint: true,
      value: originalValue,
    };
  }

  const pattern = when.pattern;
  const field = when.field;
  // what if someone's field is called '@key'?
  const isKey = field === '@key';

  if (!pattern) {
    // - if no pattern given
    //  - if field is @key THEN check if object has ANY key
    //  - else check if object[field] exists (MAY exist and be undefined!)
    if (isKey) {
      return keyAndOptionalPattern(originalValue);
    }
    return {
      lint: has(originalValue, field),
      value: originalValue,
    };
  }

  if (isKey) {
    return keyAndOptionalPattern(originalValue, pattern);
  }

  const fieldValue = String(get(originalValue, when.field));

  return {
    lint: fieldValue.match(pattern) !== null,
    value: originalValue,
  };
}

function keyAndOptionalPattern(originalValue: any, pattern?: string) {
  const type = typeof originalValue;
  switch (type) {
    case 'boolean':
    case 'string':
    case 'number':
    case 'bigint':
    case 'undefined':
    case 'function':
      return {
        lint: false,
        value: originalValue,
      };
    case 'object':
      if (originalValue === null) {
        return {
          lint: false,
          value: originalValue,
        };
      } else if (Array.isArray(originalValue)) {
        const leanValue = pattern
          ? filter(originalValue, (v, index) => {
              return String(index).match(pattern) !== null;
            })
          : originalValue;
        return {
          lint: !!leanValue.length,
          value: leanValue,
        };
      } else {
        const leanValue = pattern
          ? omitBy(originalValue, (v, key) => {
              return key.match(pattern) === null;
            })
          : originalValue;
        return {
          lint: !!Object.keys(leanValue).length,
          value: leanValue,
        };
      }
    default:
      throw new Error(
        `value: "${originalValue}" of type: "${type}" is an unsupported type of value for the "when" statement`
      );
  }
}
