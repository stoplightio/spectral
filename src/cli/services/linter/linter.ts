import { Document, STDIN } from '../../../document';
import { KNOWN_FORMATS } from '../../../formats';
import { readParsable } from '../../../fs/reader';
import * as Parsers from '../../../parsers';
import { IRuleResult, Spectral } from '../../../spectral';
import { ILintConfig } from '../../../types/config';
import { getRuleset, listFiles, skipRules } from './utils';
import { getResolver } from './utils/getResolver';

export async function lint(documents: Array<number | string>, flags: ILintConfig): Promise<IRuleResult[]> {
  const spectral = new Spectral({
    resolver: getResolver(flags.resolver),
  });

  const ruleset = await getRuleset(flags.ruleset);
  spectral.setRuleset(ruleset);

  for (const [format, lookup, prettyName] of KNOWN_FORMATS) {
    spectral.registerFormat(format, document => {
      if (lookup(document)) {
        if (!flags.quiet) {
          console.log(`${prettyName} detected`);
        }

        return true;
      }

      return false;
    });
  }

  if (flags.verbose) {
    if (ruleset) {
      const rules = Object.values(spectral.rules);
      console.info(`Found ${rules.length} rules (${rules.filter(rule => rule.enabled).length} enabled)`);
    } else {
      console.info('No rules loaded, attempting to detect document type');
    }
  }

  if (flags.skipRule) {
    spectral.setRules(skipRules(ruleset.rules, flags));
  }

  const [targetUris, unmatchedPatterns] = await listFiles(
    documents,
    !(flags.showUnmatchedGlobs || flags.failOnUnmatchedGlobs),
  );
  const results: IRuleResult[] = [];

  if (unmatchedPatterns.length > 0) {
    if (flags.failOnUnmatchedGlobs) {
      throw new Error(`Unmatched glob patterns: \`${unmatchedPatterns.join(',')}\``);
    }

    for (const unmatchedPattern of unmatchedPatterns) {
      console.log(`Glob pattern \`${unmatchedPattern}\` did not match any files`);
    }
  }

  for (const targetUri of targetUris) {
    if (flags.verbose) {
      console.info(`Linting ${targetUri}`);
    }

    const document = new Document(
      await readParsable(targetUri, { encoding: flags.encoding }),
      Parsers.Yaml,
      typeof targetUri === 'number' ? STDIN : targetUri,
    );

    results.push(
      ...(await spectral.run(document, {
        ignoreUnknownFormat: flags.ignoreUnknownFormat,
        resolve: {
          documentUri: typeof targetUri === 'number' ? void 0 : targetUri,
        },
      })),
    );
  }

  return results;
}
