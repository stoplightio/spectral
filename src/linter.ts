import { decodePointerFragment } from '@stoplight/json';
import { get } from 'lodash';
import { getDefaultRange, Resolved } from './resolved';
import { IMessageVars, message } from './rulesets/message';
import { getDiagnosticSeverity } from './rulesets/severity';
import { IFunction, IGivenNode, IRuleResult, IRunRule, IThen } from './types';
import { getClosestJsonPath, printPath, PrintStyle } from './utils';

const { JSONPath } = require('jsonpath-plus');

// TODO(SO-23): unit test but mock whatShouldBeLinted
export const lintNode = (
  node: IGivenNode,
  rule: IRunRule,
  then: IThen<string, any>,
  apply: IFunction,
  resolved: Resolved,
): IRuleResult[] => {
  const givenPath = node.path[0] === '$' ? node.path.slice(1) : node.path;
  const targetValue = node.value;

  const targets: any[] = [];
  if (then && then.field) {
    if (then.field === '@key') {
      for (const key of Object.keys(targetValue)) {
        targets.push({
          path: key,
          value: key,
        });
      }
    } else if (then.field[0] === '$') {
      try {
        JSONPath({
          path: then.field,
          json: targetValue,
          resultType: 'all',
          callback: (result: any) => {
            targets.push({
              path: JSONPath.toPathArray(result.path),
              value: result.value,
            });
          },
        });
      } catch (e) {
        console.error(e);
      }
    } else {
      // lodash lookup
      targets.push({
        path: typeof then.field === 'string' ? then.field.split('.') : then.field,
        value: get(targetValue, then.field),
      });
    }
  } else {
    targets.push({
      path: [],
      value: targetValue,
    });
  }

  if (!targets.length) {
    // must call then at least once, with no resolved
    targets.push({
      path: [],
      value: undefined,
    });
  }

  const results: IRuleResult[] = [];

  for (const target of targets) {
    const targetPath = givenPath.concat(target.path);

    const targetResults =
      apply(
        target.value,
        then.functionOptions || {},
        {
          given: givenPath,
          target: targetPath,
        },
        {
          original: node.value,
          given: node.value,
          resolved,
        },
      ) || [];

    results.push(
      ...targetResults.map<IRuleResult>(result => {
        const escapedJsonPath = (result.path || targetPath).map(segment => decodePointerFragment(String(segment)));
        const parsed = resolved.getParsedForJsonPath(escapedJsonPath);
        const path = parsed?.path || getClosestJsonPath(resolved.resolved, escapedJsonPath);
        const doc = parsed?.doc || resolved.parsed;
        const range = doc.getLocationForJsonPath(doc.parsed, path, true)?.range || getDefaultRange();
        const value = path.length === 0 ? parsed?.doc.parsed.data : get(parsed?.doc.parsed.data, path);

        const vars: IMessageVars = {
          property:
            parsed?.missingPropertyPath && parsed.missingPropertyPath.length > path.length
              ? printPath(parsed.missingPropertyPath.slice(path.length - 1), PrintStyle.Dot)
              : path.length > 0
              ? path[path.length - 1]
              : '',
          error: result.message,
          path: printPath(path, PrintStyle.EscapedPointer),
          description: rule.description,
          value,
        };

        const resultMessage = message(result.message, vars);
        vars.error = resultMessage;

        return {
          code: rule.name,
          message: (rule.message === void 0 ? rule.description ?? resultMessage : message(rule.message, vars)).trim(),
          path,
          severity: getDiagnosticSeverity(rule.severity),
          source: parsed?.doc.source,
          range,
        };
      }),
    );
  }

  return results;
};
