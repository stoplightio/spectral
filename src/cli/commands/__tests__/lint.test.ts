import * as yargs from 'yargs';

import { DiagnosticSeverity } from '@stoplight/types';
import { IRuleResult } from '../../../types';
import { lint } from '../../services/linter';
import { formatOutput, writeOutput } from '../../services/output';
import lintCommand from '../lint';

jest.mock('../../services/output');
jest.mock('../../services/linter');

function run(command: string) {
  const parser = yargs.command(lintCommand).help();
  return new Promise(done => {
    parser.parse(command, (err: Error, argv: unknown, output: string) => {
      done(output);
    });
  });
}

describe('lint', () => {
  let errorSpy: jest.SpyInstance;
  const { isTTY } = process.stdin;

  const results: IRuleResult[] = [
    {
      code: 'parser',
      message: 'Duplicate key: foo',
      path: ['foo'],
      range: {
        end: {
          character: 17,
          line: 0,
        },
        start: {
          character: 12,
          line: 0,
        },
      },
      severity: DiagnosticSeverity.Error,
    },
  ];

  beforeEach(() => {
    (lint as jest.Mock).mockClear();
    (lint as jest.Mock).mockResolvedValueOnce(results);

    (formatOutput as jest.Mock).mockClear();
    (formatOutput as jest.Mock).mockReturnValueOnce('<formatted output>');

    (writeOutput as jest.Mock).mockClear();
    (writeOutput as jest.Mock).mockResolvedValueOnce(undefined);

    errorSpy = jest.spyOn(console, 'error');
  });

  afterEach(() => {
    errorSpy.mockRestore();
    process.stdin.isTTY = isTTY;
  });

  it('shows help when no document and no STDIN are present', async () => {
    process.stdin.isTTY = true;
    const output = await run('lint');
    expect(output).toContain('documents  Location of JSON/YAML documents');
  });

  describe('when STDIN is present', () => {
    it('does not show help when documents are missing', async () => {
      const output = await run('lint');
      expect(output).not.toContain('documents  Location of JSON/YAML documents');
    });

    it('calls with lint with STDIN file descriptor', async () => {
      await run('lint');
      expect(lint).toBeCalledWith([0], {
        encoding: 'utf8',
        format: 'stylish',
        ignoreUnknownFormat: false,
        showUnmatchedGlobs: false,
        failOnUnmatchedGlobs: false,
      });
    });
  });

  it('shows help when invalid arguments are passed', async () => {
    const output = await run('lint --foo');
    expect(output).toContain('documents  Location of JSON/YAML documents. Can be either a file, a glob or');
  });

  it('calls lint with document and default options', async () => {
    const doc = './__fixtures__/empty-oas2-document.json';
    await run(`lint ${doc}`);
    expect(lint).toBeCalledWith([doc], {
      encoding: 'utf8',
      format: 'stylish',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: false,
    });
  });

  it('calls lint with document and custom encoding', async () => {
    const doc = './__fixtures__/empty-oas2-document.json';
    await run(`lint --encoding utf16 ${doc}`);
    expect(lint).toBeCalledWith([doc], {
      encoding: 'utf16',
      format: 'stylish',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: false,
    });
  });

  it('calls lint with document and custom encoding and format', async () => {
    const doc = './__fixtures__/empty-oas2-document.json';
    await run(`lint -f json --encoding utf16 ${doc}`);
    expect(lint).toBeCalledWith([doc], {
      encoding: 'utf16',
      format: 'json',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: false,
    });
  });

  it('calls lint with document and custom ruleset', async () => {
    const doc = './__fixtures__/empty-oas2-document.json';
    const ruleset = 'custom-ruleset.json';
    await run(`lint -r ${ruleset} ${doc}`);
    expect(lint).toBeCalledWith(
      [doc],
      expect.objectContaining({
        ruleset: [ruleset],
      }),
    );
  });

  it('calls lint with document and multiple custom rulesets', async () => {
    const doc = './__fixtures__/empty-oas2-document.json';
    const ruleset = 'custom-ruleset.json';
    const ruleset2 = 'custom-ruleset-2.json';
    await run(`lint --r ${ruleset} -r ${ruleset2} ${doc}`);
    expect(lint).toBeCalledWith(
      [doc],
      expect.objectContaining({
        ruleset: [ruleset, ruleset2],
      }),
    );
  });

  it.each(['json', 'stylish'])('calls formatOutput with %s format', async format => {
    await run(`lint -f ${format} ./__fixtures__/empty-oas2-document.json`);
    // needed by Node 8 (different ticking?) - can be simplified once we drop support for version 8
    await new Promise(resolve => {
      setImmediate(() => {
        expect(formatOutput).toBeCalledWith(results, format, { failSeverity: DiagnosticSeverity.Error });
        resolve();
      });
    });
  });

  it('writes formatted output to a file', async () => {
    await run(`lint -o foo.json ./__fixtures__/empty-oas2-document.json`);
    // needed by Node 8 (different ticking?) - can be simplified once we drop support for version 8
    await new Promise(resolve => {
      setImmediate(() => {
        expect(writeOutput).toBeCalledWith('<formatted output>', 'foo.json');
        resolve();
      });
    });
  });

  it('passes skip-rule to lint', async () => {
    await run('lint --skip-rule foo --skip-rule bar ./__fixtures__/empty-oas2-document.json');
    expect(lint).toHaveBeenCalledWith([expect.any(String)], {
      skipRule: ['foo', 'bar'],
      encoding: 'utf8',
      format: 'stylish',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: false,
    });
  });

  it('passes ignore-unknown-format to lint', async () => {
    await run('lint --ignore-unknown-format ./__fixtures__/empty-oas2-document.json');
    expect(lint).toHaveBeenCalledWith([expect.any(String)], {
      encoding: 'utf8',
      format: 'stylish',
      ignoreUnknownFormat: true,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: false,
    });
  });

  it('passes show-unmatched-globs to lint', async () => {
    await run('lint --show-unmatched-globs ./__fixtures__/empty-oas2-document.json');
    expect(lint).toHaveBeenCalledWith([expect.any(String)], {
      encoding: 'utf8',
      format: 'stylish',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: true,
      failOnUnmatchedGlobs: false,
    });
  });

  it('passes fail-on-unmatched-globs to lint', async () => {
    await run('lint --fail-on-unmatched-globs ./__fixtures__/empty-oas2-document.json');
    expect(lint).toHaveBeenCalledWith([expect.any(String)], {
      encoding: 'utf8',
      format: 'stylish',
      ignoreUnknownFormat: false,
      showUnmatchedGlobs: false,
      failOnUnmatchedGlobs: true,
    });
  });

  it('shows help if unknown format is passed', () => {
    return expect(run('lint -f foo ./__fixtures__/empty-oas2-document.json')).resolves.toContain(
      'documents  Location of JSON/YAML documents. Can be either a file, a glob or',
    );
  });

  it('prints error message upon exception', async () => {
    const error = new Error('Failure');
    (lint as jest.Mock).mockReset();
    (lint as jest.Mock).mockReturnValueOnce({
      // could be mockRejectedValueOnce, but Node 8 does not like it (different ticking?), so here is the workaround
      then() {
        return this;
      },
      catch(fn: Function) {
        fn(error);
        return this;
      },
    });

    await run(`lint -o foo.json ./__fixtures__/empty-oas2-document.json`);
    expect(errorSpy).toBeCalledWith('Failure');
  });
});
