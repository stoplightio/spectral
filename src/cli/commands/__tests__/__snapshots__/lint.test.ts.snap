// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint --ruleset when single ruleset option provided outputs "invalid ruleset" error 1`] = `
"Reading ruleset /home/krzysiek/development/stoplight/spectral/src/cli/commands/__tests__/__fixtures__/ruleset-invalid.yaml

/rules/rule-without-given-nor-them 	 should have required property 'given' 
/rules/rule-without-given-nor-them 	 should have required property 'then' 
/rules/rule-with-invalid-enum/severity 	 should be number 
/rules/rule-with-invalid-enum/severity 	 should be equal to one of the allowed values 
/rules/rule-with-invalid-enum/type 	 should be equal to one of the allowed values 

"
`;
